kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,958 @@
+    }
+}
+
+/// This is the callback client for the underlying physical storage driver.
+impl<'a, C: 'static + Chip> NonvolatileStorageClient for DynamicProcessLoader<'a, C> {
+    fn read_done(&self, _buffer: &'static mut [u8], _length: usize) {
+        //we will never use this, but we need to implement this anyway
+        unimplemented!();
+    }
+
+    fn write_done(&self, buffer: &'static mut [u8], length: usize) {
+        // Switch on which user generated this callback.
+        self.current_user.take().map(|user| {
+            match user {
+                NonvolatileUser::Client => {
+                    // trigger the client callback
+                    self.client.map(|client| {
+                        client.app_data_write_done(buffer, length);
+                    });
+                }
+                NonvolatileUser::Kernel => {
+                    // replace the buffer after the padding is written
+                    self.buffer.replace(buffer);
+                }
+            }
+        });
+    }
+}
+
+// Interface exposed to the app_loader capsule
+impl<'a, C: 'static + Chip> DynamicProcessLoading for DynamicProcessLoader<'a, C> {
+    fn set_client(&self, client: &'static dyn DynamicProcessLoadingClient) {
+        self.client.set(client);
+    }
+
+    fn setup(&self, app_length: usize) -> Result<usize, ErrorCode> {
+        //TODO(?): Check if it is a newer version of an existing app.
+        // We can potentially flash the new app and load it before erasing the old one.
+        // What happens to the process though? Need to delete it from the process array and load it back in?
+
+        let mut new_app_data = NewApp {
+            // struct to hold some information about the new app
+            size: 0,
+            start_addr: 0,
+        };
+
+        let flash_start = self.flash.get().as_ptr() as usize; //start of the flash region
+
+        new_app_data.start_addr = self.flash.get().as_ptr() as usize;
+        new_app_data.size = app_length;
+
+        match self.find_next_available_address(&mut new_app_data) {
+            Ok(()) => {
+                let new_start_addr = new_app_data.start_addr;
+
+                let offset = new_start_addr - flash_start;
+
+                let new_process_flash = self
+                    .flash
+                    .get()
+                    .get(offset..offset + app_length)
+                    .ok_or(ErrorCode::FAIL)?;
+
+                self.new_process_flash.set(new_process_flash);
+
+                self.new_app_start_addr.set(new_app_data.start_addr);
+                self.new_app_length.set(new_app_data.size);
+
+                // reset the struct values for a new app
+                new_app_data.size = 0;
+                new_app_data.start_addr = 0;
+
+                Ok(app_length)
+            }
+            Err(_err) => Err(ErrorCode::FAIL),
+        }
+    }
+
+    fn write_app_data(
+        &self,
+        flag: bool,
```

Comment:
`flag`?

Reply:
Fixed, this argument is now unnecessary.